WhateverGreen/kern_igfx_backlight.cpp

```diff
@@ -18,6 +18,211 @@
 /// 2. Backlight smoother that makes brightness transitions smoother on IVB+.
 ///
 
+//
+// MARK: - Backlight Registers Fix (Supplemental)
+//
+
+static constexpr const char* kHwSetBacklightSymbol = "__ZN31AppleIntelFramebufferController14hwSetBacklightEj";
+static constexpr const char* kLightUpEDPSymbol = "__ZN31AppleIntelFramebufferController10LightUpEDPEP21AppleIntelFramebufferP21AppleIntelDisplayPathPK29IODetailedTimingInformationV2";
+static constexpr const char* kHwSetPanelPowerSymbol = "__ZN31AppleIntelFramebufferController15hwSetPanelPowerEj";
+
+//
+// Supplemental Patch 1: Use WriteRegister32() to modify the register `BXT_BLC_PWM_FREQ1` instead of modifying mapped memory directly
+//
+// Function: AppleIntelFramebufferController::hwSetBacklight()
```

Comment:
AppleIntelFramebufferController::hwSetBacklight is rather short, maybe it is more reasonable to solve the field offsets and simply rewrite the function ourselves?

- We do not have Camelia, so Camelia backlight calls should not matter for us.
- We have our own logic for setting backlight based on passed backlight level (`a2` argument), so two other ifs and MMIO ops also do not matter for us.

What we are left with is:

- `mov [rbx+2E78h], r14d` backlight level set in the end of the function.
- `mov rcx, [rbx+1A08h]` MMIO space.

Perhaps do it the other way?

Reply:
> AppleIntelFramebufferController::hwSetBacklight is rather short, maybe it is more reasonable to solve the field offsets and simply rewrite the function ourselves?

I carefully revisited `hwSetBacklight` and its callers `LightUpEDP` and `hwSetPanelPower`, and I think that finding the offset of each required member field and rewriting `hwSetBacklight` is the right approach to this issue.

> We have our own logic for setting backlight based on passed backlight level, so two other ifs and MMIO ops also do not matter for us.

Yes, we do have own logic for updating registers at `0xc8254` and `0xc8258`, but I am not sure which two MMIO ops you referred to do not matter. Here is Apple's implementation and my thoughts are...

- Operations related to Camellia can be removed from our implementation.
    - i.e., No need to resolve the member field at `0x2e60`.
- Operations related to checking whether the register address is valid can be removed.
    - i.e., No need to resolve the member field at `0x1d30`.
- Operations related to fetching the base address of the MMIO region can be removed, because we will invoke the original `WriteRegister32()` explicitly.
    - i.e., No need to resolve the member field at `0x1a08`.
- Using Apple's PWM frequency leads to the 3-minute dark screen, so our implementation will use the frequency set by the firmware and thus won't write to the frequency register at 0xc8254.
    - i.e., No need to resolve the member field at `0x2e84`.
- Apple's PWM frequency divider is hard coded to `0xFFFF` and stores in the member field at `0x2e80`, but there is no guarantee that the value remains the same forever.
    - i.e., **Need to resolve the member field at `0x2e80`.**
- `hwSetPanelPower` and `LightUpEDP` always invoke `hwSetBacklight` with the brightness value stored in the member field at `0x2e78`, so it's weird that `hwSetBacklight` just stores the same value to that member field again, but I think it's better to keep this in our implementation.
     - i.e., **Need to resolve the member field at `0x2e78`.**

```c++
IOReturn AppleIntelFramebufferController::hwSetBacklight(UInt32 brightness)
{
    // Guard: Check if Camellia is present
    // The member field at 0x2e60 has a type of `CamelliaBase*`
    if (this->field_0x2e60 != nullptr)
    {
        this->field_0x2e60->SetDPCDBacklight(brightness);
    }

    // Guard: Check whether the register address `0xc8254` is valid
    // Note that this check is implemented in `WriteRegister32()`
    // The member field at 0x1d30 stores the length of the MMIO region, 
    // initialized by `AppleIntelFramebufferController::start()`
    if (0xc8254 <= this->field_0x1d30 - 4)
    {
        // The member field at 0x1a08 stores the base address of the MMIO region
        // The member field at 0x2e84 stores Apple's PWM frequency which is 
        // either 0x56CE or 0x4571 depending upon whether the raw frequency is used
        // (i.e., whether the bit 8 is set in the register `SFUSE_STRAP` at 0xc2014)
        // Set the PWM frequency
        *reinterpret_cast<UInt32*>(this->field_0x1a08 + 0xc8254) = this->field_0x2e84;
    }

    // Guard: Check whether the register address `0xc8258` is valid
    if (0xc8258 <= this->field_0x1d30 - 4)
    {
        // The member field at 0x2e80 stores Apple's PWM frequency divider
        // which is set to 0xFFFF in `AppleIntelFramebufferController::getOSInformation()`
        // Set the new duty value
        *reinterpret_cast<UInt32*>(this->field_0x1a08 + 0xc8258) = this->field_0x2e84 * brightness / this->field_0x2e80;
    }

    // Store the new brightness level to the member field at 0x2e78
    this->field_0x2e78 = brightness;

    return kIOReturnSuccess;
}
```

Your proposed approach that uses dummy controller objects is really interesting, but I think it's better to disassemble `hwSetBacklight()`, identify the offsets of the divider and the brightness level and implement the backlight registers fix in our `hwSetBacklight()` wrapper. Now the problem lies in the caller site. I managed to disassemble those two callers and generate the assembly patch for each of them at runtime so that they will call `hwSetBacklight()` explicitly. I have added a new patch submodule to reflect the new changes and can confirm that it works on my machine properly. It would be great if you can review it and provide suggestions on making the new submodule easier to maintain in future. Once you approve the new changes, I will remove the supplemental fix from the WEG.